linera-service/src/chain_listener.rs

```diff
@@ -43,10 +43,16 @@ pub struct ChainListenerConfig {
 }
 
 #[async_trait]
-pub trait ClientContext<P: ValidatorNodeProvider> {
+pub trait ClientContext {
+    type ValidatorNodeProvider: LocalValidatorNodeProvider;
+    type Storage: Storage;
```

Comment:
Could these be two type parameters? (similar to `ChainClient`)

Reply:
They can be, but I think they shouldn't be.  Having them as trait _parameters_ implies that the same `ClientContext` instance can be used for different types, e.g. we could create, from the same client, two different chain clients for the same chain with different storage backends and validator nodes.